CONTRIBUTING.md

```diff
@@ -0,0 +1,439 @@
+- Ternary operator use only within a single line total.
+
+Fabric API uses the Gradle checkstyle plugin. If you run `gradle check` any style errors will be logged in the output. The style errors can also be viewed in a generated webpage by the Gradle task.
+
+### Style guidelines
+
+These are less strict guidelines for the code style, but you should generally obey these:
+
+- No redundant static access qualifiers.
+  - For example if you call `ClassHere.staticThing()` inside of another method in `ClassHere`, you should avoid fully qualifying the static access to `staticThing()`.
+- The use of the `this` qualifier is optional. You may need to use it in order to avoid shadowing variables but its general use elsewhere is fine.
+  - Generally keep consistent with the current standard inside of the class you are editing.
+
+## Testing
+Including testing code when submitting new features is essential, both to demonstrate the feature, and to ensure that it works correctly.
+Testing code should not be in the regular source sets, it belongs to so-called "test mods".
+
+- Test mods are located in the `testmod` source set. Example: `fabric-lifecycle-events-v1/src/testmod/`
+- Test mods should provide the necessary content to debug an issue with the api manually.
+  - Test mods can also be helpful in order to illustrate how to use a module as a fallback for wiki pages that have yet to be written.
+- If a test mod can be partially automated then it is encouraged to implement an automatically failing test if something goes wrong.
+  - This allows issues with the implementation or porting issues to be detected immediately.  
+  - For example, if a test mod checks if commands are registered on the server and the commands are not on the `CommandDispatcher` then the test should fail. 
+  - Good places to run the automated checks include:
+    - The mod initializer if applicable.
+    - In a listener for `ServerLifecycleEvents.SERVER_STARTED` if a server instance is required.
+    - A game test.
+  - Test failures should always throw an `AssertionError` explaining what condition was not met during the test.
+
+Fabric API pull requests should be tested in the dev environment and in production (on both a client and dedicated server).
+The `gradlew build` command can be used to produce the Fabric API fatjar, located in `builds/libs/`.
+
+#### Common mistakes
+One highly likely cause of a production failure is the use of `remap=false` in a mixin. If `remap=false` is used, you need to verify the mixin works in dev and production. Most likely the mixin will not work in production.
+
+## Checklist before submitting a pull request
+
+### Apply license headers
+There is a Gradle task that can automate this for you. Simply run `gradlew spotlessApply`.
+
+### Run checks
+- The `gradlew check` task runs all the style checks and game tests.
+- Code style can be checked with the faster `gradlew checkstyleMain checkstyleTestmod`.
```

Comment:
Maybe create a new task that run all the checkstyle tasks?
```gradle
task checkstyle {
	dependsOn tasks.withType(Checkstyle)
	subprojects.forEach {
		dependsOn it.tasks.withType(Checkstyle)
	}
}
```

Reply:
Yeah, something along those lines sounds reasonable.